### PR TITLE
[Benchmark] Add a multi-round QA/chat performance benchmark  

### PR DESCRIPTION
**Pull request by ApostaC**
This PR contains a multi-round QA benchmark for users to run.

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to LMCache! Before submitting the pull request, please ensure the PR meets the following criteria. This helps us maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Please try to classify PRs for easy understanding of the type of changes. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Core]</code> for changes in the core LMCache logic (e.g., <code>LMCacheEngine</code>, <code>Backend</code> etc.)</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>Pass all linter checks. Please use <a href="https://github.com/LMCache/LMCache/blob/dev/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li> Please include sufficient tests to ensure the change is stay correct and robust. This includes both unit tests and integration tests.</li>
</ul>

<h3>What to Expect for the Reviews</h3>

We aim to address all PRs in a timely manner. If no one reviews your PR within 5 days, please @-mention one of KuntaiDu, ApostaC or YaoJiayi.

### CODE REVIEW

**Comment by Hanchenli** [comment]
Can you add some brief description of the workload?

**Comment by ApostaC** [comment]
> Can you add some brief description of the workload?

@Hanchenli Just added readme

File: benchmarks/utils.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        def stop_loop():
+            cls._logger.debug("Stoping the loop!")
+            cls._loop.stop()
+
+
+        cls._logger.info("Waiting for remaining tasks to finish")
+        cls.WaitLoop()
+        #async def wait_for_tasks():
+        #    current_task = asyncio.current_task(cls._loop)
+        #    tasks = [task for task in asyncio.all_tasks(cls._loop) if not task.done() and task is not current_task]
```
Comment (Hanchenli) [inline benchmarks/utils.py]: Should we remove these?

File: benchmarks/multi-round-qa.py [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+            )
+            for session in sessions_to_remove:
+                self.session_summaries.append(session.summary())
+        self.sessions = [s for s in self.sessions if not s.finished]
+
+    def step(self, timestamp: float, executor: RequestExecutor):
+        if self.start_time is None:
+            self.start_time = timestamp
+
+        if timestamp - self.last_user_join > self.gap_between_users:
```
Comment (Hanchenli) [inline benchmarks/multi-round-qa.py:338]: Maybe specify that the gap is now fixed and users do not overlap.

File: benchmarks/README.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
+# Benchmarking LLM Performance: Multi-Round QA Use Case
+
+## Overview
+
+This repository contains benchmarking tools for evaluating the performance of language models in various scenarios. The initial focus of this benchmark is on the multi-round QA (Question Answering) use case. The script `multi-round-qa.py` simulates multiple users interacting with a language model concurrently, allowing you to analyze the model's response quality, throughput, and latency.
```
Comment (Hanchenli) [inline benchmarks/README.md]: We should add a comparison with baseline vLLM in every performance comparison as well.

**Review by YaoJiayi — COMMENTED** [review]
A quick question: why are we not puting this in the lmcache-test repo? Isn't most code resuable there?

**Comment by ApostaC** [comment]
> A quick question: why are we not puting this in the lmcache-test repo? Isn't most code resuable there?

Good point. The goal of this is for users to run something themselves. lmcache-tests repo is for our internal use and it's pretty complicated.

Another thing is this script only has the "query sender" without any serving engine launching script. In this case, anybody can launch any serving engine themselves (as long as it's openai compatible)

File: benchmarks/README.md (hunk 3 above)
Comment (ApostaC) [reply benchmarks/README.md]: Do you mean we should post some figures at somewhere periodically?

File: benchmarks/multi-round-qa.py (hunk 2 above)
Comment (ApostaC) [reply benchmarks/multi-round-qa.py:338]: Users do overlap. It's more like:
```
--------------------------------------------------------> TIME
User 1: req1    req2    req3    req4    req5
User 2:      req1    req2    req3    req4    req5
User 3:            req1    req2    req3    req4    req5
User 4:                  req1    req2    req3    req4    req5
```

**Comment by YaoJiayi** [comment]
> > A quick question: why are we not puting this in the lmcache-test repo? Isn't most code resuable there?
> 
> Good point. The goal of this is for users to run something themselves. lmcache-tests repo is for our internal use and it's pretty complicated.
> 
> Another thing is this script only has the "query sender" without any serving engine launching script. In this case, anybody can launch any serving engine themselves (as long as it's openai compatible)

Makes sense.

**Review by YaoJiayi — APPROVED** [review]

**Pull request merged**